File: AkkaWithZeroMQ/src/main/scala/org/akka/essentials/zeromq/example2/RouterActor.scala
package org.akka.essentials.zeromq.example2
import akka.actor.actorRef2Scala
import akka.actor.Actor
import akka.actor.ActorLogging
import akka.util.duration.intToDurationInt
import akka.zeromq.Bind
import akka.zeromq.Frame
import akka.zeromq.HighWatermark
import akka.zeromq.SocketType
import akka.zeromq.ZMQMessage
import akka.zeromq.ZeroMQExtension
import scala.util.Random
import akka.zeromq.Listener
import akka.actor.Cancellable

case class Tick

class RouterActor extends Actor with ActorLogging {
  val pubSocket = ZeroMQExtension(context.system).newSocket(SocketType.Router, Bind("tcp://127.0.0.1:1234"), Listener(self), HighWatermark(50000))

  var random = new Random(3);

  var count = 0
  var cancellable: Cancellable = null
  override def preStart() {
    cancellable = context.system.scheduler.schedule(1 second, 1 second, self, Tick)
  }
  def receive: Receive = {
    case Tick =>
      count += 1
      if (random.nextBoolean() == true) {
        pubSocket ! ZMQMessage(Seq(Frame("A"), Frame("This is the workload for A")))
      } else {
        pubSocket ! ZMQMessage(Seq(Frame("B"), Frame("This is the workload for B")))
      }
      if (count == 10) {
        cancellable.cancel()
      }

    case m: ZMQMessage =>
      var replier = new String(m.payload(0))
      var msg = new String(m.payload(1))
      log.info("recived message from {} with mesg -> {}", replier, msg)
  }
}

File: AkkaWithZeroMQ/src/main/java/org/akka/essentials/zeromq/example2/RouterActor.java
package org.akka.essentials.zeromq.example2;

import java.util.Random;

import akka.actor.ActorRef;
import akka.actor.Cancellable;
import akka.actor.UntypedActor;
import akka.event.Logging;
import akka.event.LoggingAdapter;
import akka.util.Duration;
import akka.zeromq.Bind;
import akka.zeromq.Frame;
import akka.zeromq.HighWatermark;
import akka.zeromq.Listener;
import akka.zeromq.SocketOption;
import akka.zeromq.ZMQMessage;
import akka.zeromq.ZeroMQExtension;

public class RouterActor extends UntypedActor {
	public static final Object TICK = "TICK";

	Random random = new Random(3);
	int count = 0;
	Cancellable cancellable;

	ActorRef routerSocket = ZeroMQExtension.get(getContext().system())
			.newRouterSocket(
					new SocketOption[] { new Listener(getSelf()),
							new Bind("tcp://127.0.0.1:1237"),
							new HighWatermark(50000) });
	LoggingAdapter log = Logging.getLogger(getContext().system(), this);

	@Override
	public void preStart() {
		cancellable = getContext()
				.system()
				.scheduler()
				.schedule(Duration.parse("1 second"),
						Duration.parse("1 second"), getSelf(), TICK);
	}

	@Override
	public void onReceive(Object message) throws Exception {
		if (message.equals(TICK)) {

			if (random.nextBoolean() == true) {
				routerSocket.tell(new ZMQMessage(new Frame("A"), new Frame(
						"This is the workload for A")));
			} else {
				routerSocket.tell(new ZMQMessage(new Frame("B"), new Frame(
						"This is the workload for B")));

			}
			count++;
			if (count == 10)
				cancellable.cancel();

		} else if (message instanceof ZMQMessage) {
			ZMQMessage m = (ZMQMessage) message;
			String replier = new String(m.payload(0));
			String msg = new String(m.payload(1));
			log.info("Received message from {} with mesg -> {}", replier, msg);
		}

	}

}
